### PR TITLE
Restore readonly DimCoords

### PR DESCRIPTION
**Pull request by djkirkham**
Historically it's always been the case that the points and bounds arrays of a DimCoord haven't been modifiable from outside the class. Recently this has slipped.

There are two changes in this PR. ~One I consider an obvious bug: when converting the argument to the points/bounds setter to a numpy array, a flag is created which determines whether to make a copy. Currently this flag is set if the input is lazy, but this is the wrong way around. If the input argument is already a numpy array, the points will be modifiable via the original array passed in. Conversely, if a lazy array is passed in then we can safely use the array returned by `as_concrete_data` as the actual points array, since there's no way a user could have a reference to it.~ Edit: I've realised that even if the input array is lazy we need to copy the result of `as_concrete_data`, since it could simply be a dask wrapper around a NumPy array to which the user has access.

The other change might be considered more controversial. Although the points array returned by the points/bounds array is read-only, it is not a view. This means it's possible for a user to change metadata like dtype and shape. I've changed it so that the getter always returns a view.

### CODE REVIEW

File: lib/iris/coords.py [hunk 1]
```diff
@@ -1677,13 +1677,15 @@ def _new_points_requirements(self, points):
         if points.size > 1 and not iris.util.monotonic(points, strict=True):
             raise ValueError('The points array must be strictly monotonic.')
 
+    def _points_getter(self):
+        return super(DimCoord, self)._points_getter().view()
```
Comment (pelson) [inline lib/iris/coords.py]: Why aren't we doing this for AuxCoords too?

**Comment by pelson** [comment]
I'm OK with it being a view, but wonder why this only applies to DimCoords. Why not AuxCoords?

File: lib/iris/coords.py (hunk 1 above)
Comment (djkirkham) [reply lib/iris/coords.py]: Makes sense. I'll change it.

File: lib/iris/coords.py [hunk 2]
```diff
@@ -1761,6 +1759,16 @@ def _bounds_setter(self, bounds):
 
     bounds = property(Coord._bounds_getter, _bounds_setter)
 
+    def core_points(self):
```
Comment (pelson) [inline lib/iris/coords.py]: Move into the Coord?

**Comment by djkirkham** [comment]
I'm about to push another change, please don't merge this until then.

File: lib/iris/_data_manager.py [hunk 3]
```diff
@@ -226,8 +226,8 @@ def data(self):
                         'before trying again.')
                 raise MemoryError(emsg.format(self.shape, self.dtype))
 
-        # Check the manager contract, as the managed data has changed.
-        self._assert_axioms()
+            # Check the manager contract, as the managed data has changed.
+            self._assert_axioms()
```
Comment (djkirkham) [inline lib/iris/_data_manager.py]: I've indented this because we only need to assert the axioms if the state changed.

**Pull request merged**